parsers/yaml/yaml.cc

```diff
@@ -0,0 +1,2776 @@
+            (T(EmptyLine, WhitespaceLine) * End) >>
+          [](Match&) -> Node { return nullptr; },
+
+        In(MappingItem, SequenceItem) *
+            (T(KeyGroup, ValueGroup)[Group]
+             << (T(Line) << (T(Whitespace) * End))) >>
+          [](Match& _) {
+            Node group = _(Group);
+            group->erase(group->begin(), group->begin() + 1);
+            return group;
+          },
+
+        In(ComplexKey) * (T(Hyphen) * ValueToken[Value]) >>
+          [](Match& _) { return SequenceIndent << (SequenceItem << _(Value)); },
+
+        In(ComplexKey, ComplexValue) * (T(Line) << End) >>
+          [](Match&) -> Node { return nullptr; },
+
+        In(ComplexKey, ComplexValue) *
+            (T(Line) << (AnchorTag++[Anchor] * T(Literal, Folded)[Block])) >>
+          [](Match& _) {
+            Node anchortag = Line << _[Anchor];
+            if (anchortag->empty())
+            {
+              return BlockStart << _(Block);
+            }
+
+            return Seq << anchortag << (BlockStart << _(Block));
+          },
+
+        In(ComplexKey) * (AnchorTag[Anchor] * End) >>
+          [](Match& _) { return Seq << _(Anchor) << (Empty ^ ""); },
+
+        In(ComplexValue) * (AnchorTag[Anchor] * End) >>
+          [](Match& _) { return Seq << _(Anchor) << (Null ^ "null"); },
+
+        In(DocumentGroup, KeyGroup, ValueGroup) * ((T(Indent, Line)) << End) >>
+          [](Match&) -> Node { return nullptr; },
+
+        In(MappingGroup, SequenceGroup) *
+            (T(Line) << (~T(Whitespace) * T(Comment))) >>
+          [](Match&) -> Node { return nullptr; },
+
+        In(MappingGroup) * (T(ComplexValue) << End) >>
+          [](Match&) { return ComplexValue << (Null ^ "null"); },
+
+        In(ComplexKey, ComplexValue) *
+            ((T(Indent) << (T(Line) << (~T(Whitespace) * T(Comment)))) * End) >>
+          [](Match&) -> Node { return nullptr; },
+
+        In(SequenceGroup, MappingGroup) * T(WhitespaceLine, EmptyLine) >>
+          [](Match&) -> Node { return nullptr; },
+
+        In(Documents) *
+            (T(Document) << (T(Directives) << End) *
+               (T(DocumentGroup) << End)) >>
+          [](Match&) -> Node { return nullptr; },
+
+        In(MappingIndent) *
+            (T(MappingGroup)[Group]
+             << (T(MappingItem, ComplexKey, ComplexValue)++ * End)) >>
+          [](Match& _) { return Seq << *_[Group]; },
+
+        In(SequenceIndent) *
+            (T(SequenceGroup)[Group] << (T(SequenceItem)++ * End)) >>
+          [](Match& _) { return Seq << *_[Group]; },
+
+        // errors
+
+        In(BlockStart) * BasicToken[Value] >> [](Match& _) -> Node {
+          return err(_(Value), "Invalid text after block scalar indicator");
+        },
+
+        In(SequenceItem) *
+            (T(ValueGroup) << (T(FlowMapping, FlowSequence))[Flow])(
+              [](auto& n) {
+                Node group = n.first[0];
+                Node item = group->parent()->shared_from_this();
```

Comment:
I think Trieste should suppport a parent on NodeRange.  This seems to be a common pattern in your code.